### PR TITLE
Basic firebase integration.

### PR DESCRIPTION
**Pull request by shans**
This still needs some work on correct versioning, and I haven't checked whether
it does events correctly yet, but it's complete enough to warrant a review.

### CODE REVIEW

File: runtime/firebase-storage.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+      this._fire('change', {data: this._setToList(this.dataSnapshot.val()), version: this._version});
+    });
+  }
+
+  async get(id) {
+    return this.dataSnapshot.val()[id];
+  }
+
+  async store(entity) { 
+    return this.reference.child(entity.id).set(entity);
```
Comment (dstockwell) [inline runtime/firebase-storage.js]: call `remove()` instead of set if entity is null

File: runtime/firebase-storage.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    if (this.databaseUrl && this.apiKey)
+      return `${this.protocol}://${this.databaseUrl}/${this.apiKey}`;
+    return `${this.protocol}://`;
+  }
+}
+
+export class FirebaseStorage {
+  constructor(arc) {
+    this._arc = arc;
+    this.apps = {};
```
Comment (dstockwell) [inline runtime/firebase-storage.js]: why not `_apps` etc?

File: runtime/firebase-storage.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  toString() {
+    if (this.databaseUrl && this.apiKey)
+      return `${this.protocol}://${this.databaseUrl}/${this.apiKey}`;
+    return `${this.protocol}://`;
+  }
+}
+
+export class FirebaseStorage {
```
Comment (dstockwell) [inline runtime/firebase-storage.js]: Are we planning to export other things, otherwise just export default.

File: runtime/firebase-storage.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  }
+}
+
+class FirebaseCollection extends FirebaseStorageProvider {
+  constructor(type, arc, id, firebaseApp, firebaseKey) {
+    super(type, arc, id, firebaseApp, firebaseKey);
+    this.dataSnapshot = undefined;
+    this.reference.on('value', dataSnapshot => {
+      this.dataSnapshot = dataSnapshot;
+      this._fire('change', {data: this._setToList(this.dataSnapshot.val()), version: this._version});
```
Comment (dstockwell) [inline runtime/firebase-storage.js]: TODO about changing this to use child_added/removed? I think we'll want to be able to subscribe to similar events in the particle too.

File: runtime/firebase-storage.js (hunk 3 above)
Comment (shans) [reply runtime/firebase-storage.js]: Done.

File: runtime/firebase-storage.js (hunk 2 above)
Comment (shans) [reply runtime/firebase-storage.js]: done.

File: runtime/firebase-storage.js (hunk 1 above)
Comment (shans) [reply runtime/firebase-storage.js]: if entity is null there's nothing to remove. We probably need to augment the collection API with a remove(id) method instead.

File: runtime/firebase-storage.js (hunk 4 above)
Comment (shans) [reply runtime/firebase-storage.js]: We probably want to do a round of design in this area, particularly as we don't know what the exact data transfer approach used by firebase is.

**Pull request merged**